### ISSUE TITLE
node < 4 support PR?

### ISSUE THREAD

**Issue by olsonpm** [patrick-steele-idem/child-process-promise#21]
Would you accept a pr which utilizes webpack + babel to expose an es5 version?

Currently the line lower versions of node trip up on lib/index.js:139

```js
var err = {
  code: code,
  message: '`' + commandStr + '` failed with code ' + code,
  childProcess: cp,
  toString() { // <------------- invalid syntax
    return this.message;
  }
};
```

You could keep the default require pointing to your current entry point, and have webpack output a file `es5.js` which would allow people to `require('child-process-promise/es5')`.

As for "why?" - my corporate environment runs build servers using node v0.12.14 and so our development environment should run on the same.  Getting my co-workers using the command line is hard enough, but if they have to remember to mess with nvm in order to use the few CLI tools I create, that's a hurdle they shouldn't have to deal with.

I appreciate the module regardless!

**Comment by patrick-steele-idem** [comment]
I would be okay with integrating babel for transpiling and generating an ES5 build as a `prepublish` step, but I see no need for webpack. Invoking the `babel` command line utility seems like the easiest option:

```bash
babel src --out-dir src-es5
```

In addition, I would be okay with providing a default Promise polyfill (which would only be loaded if `typeof Promise === 'undefined'`) and I think we should automatically export the ES5 or ES6 version based on the Node.js runtime.

What do you think?

**Comment by olsonpm** [comment]
Ha - all that sounds awesome.  I only mentioned webpack because it's the process I've used in personal projects.  I trust your knowledge on the cli being simpler, as I find webpack to be a confusing hack of configurations.

I haven't messed with implementing a promise polyfill before but am confident I can figure it out as I've read a decent amount.  I've just always had bluebird at my disposal.

> and I think we should automatically export the ES5 or ES6 version based on the Node.js runtime.

you mean something like:

```js
var es5 = require('./src-es5')
  , es6 = require('./src');

return (isGTEV4(process.version))
  ? es6
  : es5;
```

?

So basically I'll have to read more about best practices - for some reason requiring both sources (or conditionally requiring) sounds odd to me.

Keep in mind I'm no veteran node dev ha

**Comment by patrick-steele-idem** [comment]
To avoid loading both versions you would want to do the following:

```javascript
return isES6Supported() ?
    require('./src') :
    require('./src-es5');
```

For determining the version, you could use the following module:
https://github.com/srod/node-version

Or simply:

```javascript
var majorVersion = parseInt(process.versions.node.split('.')[0], 10)
```

For the `Promise` polyfill I would use the following lib: https://github.com/taylorhakes/promise-polyfill

Thanks for looking into this.

**Comment by olsonpm** [comment]
I see - the conditional require problem is specific to packers and would require workarounds.  [This thread](https://github.com/substack/node-browserify/issues/664) explains why it's an issue for browserify, and there are various plugins for both browserify and webpack to address the issue (e.g. configure the packager with an environment variable it can then use to conditionally require).

So then it just depends on whether that use-case matters to you.

I will try to have something by this weekend.  I appreciate your prompt feedback.

**Comment by patrick-steele-idem** [comment]
> the conditional require problem is specific to packagers and would require workarounds.

There should be no need to send `child-process-promise` through a JavaScript module bundler. I'm not worried about that use case since this code only runs on the server.

As a somewhat related side note, [lasso](https://github.com/lasso-js/lasso) handles conditional/dynamic requires just fine since it faithfully recreates the Node.js module loader in the browser and supports dynamic and conditional loading of JavaScript modules.

**Comment by olsonpm** [comment]
Interesting - yeah I thought there were non-browser use-cases for bundlers, but the more I google the more I find "how to bundle for the server environment" instead of "why".  I'll use the conditional as you explained.

Thanks for mentioning lasso - hadn't heard of it before.

**Comment by olsonpm** [comment]
Working on it now and will create a PR.  I should have closed this issue when we resolved the direction anyway.
